lib/hanami/container.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+require "dry/system/container"
+
+module Hanami
+  # Hanami private IoC
+  #
+  # @since 2.0.0
+  class Container < Dry::System::Container
+    boot(:root) do
+      start do
+        register(:root, Hanami.root)
+      end
+    end
+
+    boot(:env) do |c|
+      init do
+        use :root
+
+        begin
+          require "dotenv"
+        rescue LoadError # rubocop:disable Lint/HandleExceptions
+        end
+
+        Dotenv.load(c[:root].join(".env")) if defined?(Dotenv)
+      end
+    end
+
+    boot(:lib) do |c|
+      init do
+        use :root
+
+        $LOAD_PATH.unshift root.join("lib")
+        Hanami::Utils.require!(c[:root].join("lib", "**", "*.rb"))
+      end
+    end
+
+    boot(:configuration) do |c|
+      init do
+        use :root
+
+        require c[:root].join("config", "application").to_s
+      end
+
+      start do
+        register(:configuration, Hanami.application.configuration.finalize)
+      end
+    end
+
+    boot(:logger) do |c|
+      init do
+        require "hanami/logger"
+      end
+
+      start do
+        use :configuration
+        register(:logger, Hanami::Logger.new(c[:configuration].logger))
+      end
+    end
+
+    boot(:routes) do |c|
+      init do
+        require c[:root].join("config", "routes").to_s
+      end
+
+      start do
+        register(:routes, Hanami.application.routes)
+      end
+    end
+
+    boot(:apps) do |c|
+      start do
+        use :routes
+
+        register(:apps, c[:routes].apps)
+      end
+    end
+
+    boot(:actions) do |c|
+      init do
+        use :configuration
+        use :apps
+
+        c[:apps].each do |app|
+          require c[:root].join("apps", app.to_s, "action")
+
+          namespace = Utils::String.classify("#{app}::Actions")
+          namespace = Utils::Class.load!(namespace)
+          # action    = Utils::Class.load!("#{namespace}::Action")
+          configuration = Controller::Configuration.new do |config|
+            config.cookies                 = c[:configuration].cookies.options
+            config.default_headers         = c[:configuration].security.to_hash
+            config.default_request_format  = c[:configuration].default_request_format
+            config.default_response_format = c[:configuration].default_response_format
+          end
+
+          register(:"apps.#{app}.actions.namespace", namespace)
+          register(:"apps.#{app}.actions.configuration", configuration)
+        end
+      end
+    end
+
+    boot(:code) do |c|
+      init do
+        use :root
+        use :configuration
+        use :apps
+        use :actions
+
+        apps = c[:apps].join(",")
+        Hanami::Utils.require!(c[:root].join("apps", "{#{apps}}", "**", "*.rb"))
+      end
+    end
```

Comment:
@timriley This sounds like a nice to have feature, but without a real, immediate demand for it. I'd pass it for now.

Reply:
@jodosha I'm fine if you want to pass it for now (i.e. for this PR), but I think we should have a discussion about this before 2.0 settles in full. From actual practice (i.e. several quite large apps here at Icelab), having separate containers turns out to be a really helpful way to arrange/separate the components for distinct sections (i.e. slices) within a larger application and you're probably going to have to reimplement some of the resulting behaviour yourself when you work out selective booting of the slices.

Perhaps we can discuss sometime (along with @solnic or @flash-gordon) after this PR is merged?

Reply:
Yes, having multiple containers per slice and giving them access to the "main" container so that they can import core components (if needed) is a great architecture that was already proven to work well in practice. Making it work like that from the beginning will certainly reduce the risk of having to do bigger'ish refactorings later on. It doesn't have to be a part of this PR but it'd be great to address this prior 2.0 final. I'm happy to help with this.

Reply:
Blah sorry, I meant *having multiple containers per app where each slice has its own container*, of course 😄

Reply:
Oh, that's interesting. This doesn't belong really to this PR. We'll need to discuss about this.